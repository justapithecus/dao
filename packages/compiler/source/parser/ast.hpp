#pragma once

#include <memory>
#include <string>
#include <variant>
#include <vector>

namespace dao {

  struct number_expr;
  struct identifier_expr;
  struct binary_expr;
  struct call_expr;
  struct prototype;
  struct function_def;

  using ast_node = std::variant<number_expr, identifier_expr>;

  // using ast_node = std::variant<number_expr, variable_expr, binary_expr, call_expr,
  //   prototype, function_def>;

  struct number_expr {
    int val;
  };

  struct identifier_expr {
    std::string name;
  };

  // struct binary_expr {
  //   char                      op;
  //   std::unique_ptr<ast_node> lhs;
  //   std::unique_ptr<ast_node> rhs;
  // };

  // struct call_expr {
  //   std::string                            callee;
  //   std::vector<std::unique_ptr<ast_node>> args;
  // };

  // struct prototype {
  //   std::string              name;
  //   std::vector<std::string> args;
  // };

  // struct function_def {
  //   std::unique_ptr<prototype> proto;
  //   std::unique_ptr<ast_node>  body;
  // };

  static_assert(sizeof(ast_node) <= 56);

} // namespace dao
