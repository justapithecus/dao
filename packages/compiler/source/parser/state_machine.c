#include "state_machine.h"

#define offset(kind) ((kind) * (lexical_state_count))

unsigned char const ch_to_eqc[256] = {
  // Layout
  [0]          = offset(glyph_eof),
  [1 ... 9]    = offset(glyph_layout),
  ['\n']       = offset(glyph_new_line),
  ['\r']       = offset(glyph_new_line),
  [11 ... 12]  = offset(glyph_layout),
  [14 ... 31]  = offset(glyph_layout),
  [' ']        = offset(glyph_white_space),
  // Special
  ['!']        = offset(glyph_operator),
  ['(']        = offset(glyph_operator),
  [')']        = offset(glyph_operator),
  ['*']        = offset(glyph_operator),
  ['+']        = offset(glyph_operator),
  ['-']        = offset(glyph_operator),
  ['.']        = offset(glyph_operator),
  ['/']        = offset(glyph_operator),
  ['<']        = offset(glyph_operator),
  ['=']        = offset(glyph_operator),
  ['>']        = offset(glyph_operator),
  // Numbers
  [48 ... 57]  = offset(glyph_number),
  // Letters
  [65 ... 90]  = offset(glyph_letter),
  ['_']        = offset(glyph_letter),
  [97 ... 122] = offset(glyph_letter),
};
