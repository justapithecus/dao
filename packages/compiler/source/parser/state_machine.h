#pragma once

// Glyphs are categories comprising individual characters or strings of characters, e.g.
// '+' and '->', respectively.
typedef enum glyph {
  glyph_layout,
  glyph_white_space,
  glyph_new_line,
  glyph_letter,
  glyph_number,
  glyph_separator,
  glyph_operator,
  glyph_eof,
  glyph_count,
} glyph;

typedef enum lexical_state {
  lexical_state_final,
  lexical_state_next_char,
  lexical_state_separator,
  lexical_state_identifier,
  lexical_state_identifier_end,
  lexical_state_numeral,
  lexical_state_numeral_end,
  lexical_state_operator,
  lexical_state_new_line,
  lexical_state_eof,
  lexical_state_error,
  lexical_state_count,
} lexical_state;

// Character to character-equivalence classes lookup table
//
// Character-equivalence classes are pre-computed offsets used to index into
// the lexical state transitions lookup table.
extern unsigned char const equivalence_class[256];

// Lexical state transitions lookup table
//
// Index using the current state offset by a a character-equivalence class, to lookup
// the next lexical state.
#define LEX_TRANS_SIZE ((glyph_count) * (lexical_state_count))
#pragma clang diagnostic   push
#pragma clang diagnostic   ignored "-Wdeprecated-enum-enum-conversion"
extern unsigned char const transition[LEX_TRANS_SIZE];
#pragma clang diagnostic   pop

// Character rewind
//
// Certain lexical states, typically those terminating a lexeme, requires shifting back 1
// character, to start lexing anew.
extern char const ch_rewind[lexical_state_count];

// True if inside an identifier, value, or operator
extern unsigned char const inside[lexical_state_count];

extern unsigned char const binary_op_precedence[256];
