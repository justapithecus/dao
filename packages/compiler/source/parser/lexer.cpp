#include "../utils.hpp"

#include "lexer.hpp"
#include "state_machine.h"

namespace dao {

  auto lex(std::string_view fpath) -> std::vector<dao::token> {
    std::vector<dao::token> tokens{};
    dao::lexeme             lexeme{};
    auto                    source{dao::fread(fpath.data())};
    auto                    src_ptr{source.data()};
    unsigned char           state{lexical_state_next_char};

    do {
      unsigned char ch;
      auto          eqc = equivalence_class[(ch = *src_ptr++)];

      state = transition[state + eqc];
      src_ptr += ch_rewind[state];
      // lexeme.col_num += ch_rewind[state];
      lexeme.len += inside[state];

      switch (state) {
      case lexical_state_next_char: {
        ++lexeme.col_num;
        break;
      }
      case lexical_state_identifier: {
        lexeme.token_kind = token_kind_identifier;
        break;
      }
      case lexical_state_numeral: {
        lexeme.token_kind = token_kind_numeral;
        break;
      }
      case lexical_state_operator:
        lexeme.token_kind = token_kind_operator;
        [[fallthrough]];
      case lexical_state_identifier_end:
        [[fallthrough]];
      case lexical_state_numeral_end: {
        lexeme.update_repr(src_ptr, 0);
        lexeme.reset();
        tokens.emplace_back(lexeme.as_token());
        break;
      }
      default:
        break;
      }

    } while (state > lexical_state_final);

    return tokens;
  }
} // namespace dao
