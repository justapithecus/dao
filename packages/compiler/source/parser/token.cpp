#include <iomanip>

#include "state_machine.h"
#include "token.hpp"

#define token_desc(kind, desc)                                                 \
  case (kind):                                                                 \
    os << (desc);                                                              \
    break

namespace dao {

  auto operator<<(std::ostream &os, token const &tok) -> std::ostream & {
    os << std::setw(12);

    switch (tok.kind) {
      token_desc(token_kind_identifier, "Identifier");
      token_desc(token_kind_numeral, "Numeral");
      token_desc(token_kind_operator, "Operator");
      token_desc(token_kind_separator, "Separator");
    default:
      os << "Unknown";
      break;
    }

    return os << " | " << tok.repr << " |";
  }
} // namespace dao
