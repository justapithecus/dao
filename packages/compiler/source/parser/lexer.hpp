#pragma once

#include <iostream>
#include <string_view>
#include <vector>

#include "token.hpp"

namespace dao {

  struct lexeme {
    std::string  repr;
    std::uint8_t len;
    std::uint8_t line_num;
    std::uint8_t col_num;

    explicit lexeme()
      : len{0}
      , line_num{1}
      , col_num{1} {
    }

    auto update_repr(char const *src_ptr, int end_offset) {
      auto first{src_ptr - len};
      auto last{first + (len + end_offset)};
      repr = std::string{first, last};
    }

    // reset trackers for next lexeme
    auto reset(int offset = 0) {
      col_num += (len + offset);
      len = 0;
    }

    auto as_token(dao::token_kind kind) {
      reset();
      return dao::token{repr, kind};
    }
  };

  static_assert(sizeof(dao::lexeme) <= 32);

  auto lex(std::string_view fpath) -> std::vector<token>;
} // namespace dao
