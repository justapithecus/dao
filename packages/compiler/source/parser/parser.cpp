#include "parser/parser.hpp"
#include "ast.hpp"
#include "state_machine.h"
#include "token.hpp"

namespace dao {

  auto parse(std::vector<dao::token> const &tokens) -> std::unique_ptr<ast_node> {
    parse_context ctx{tokens};

    for (auto tok{ctx.peek()}; not ctx.is_eof();) {

      switch (tok->kind) {
      case token_kind_identifier:
        return parse_identifier_expr(ctx);
      }
    }

    return nullptr;
  }

  auto parse_identifier_expr(parse_context &ctx) -> std::unique_ptr<dao::ast_node> {
    identifier_expr expr{ctx.peek()->repr};
    ctx.eat();
    return std::make_unique<dao::ast_node>(std::move(expr));
  }

  // auto parse_number_expr(cursor *tok) {
  //   auto val = std::stoi(tok->base()->repr);
  //   tok++;
  //   return std::make_unique<ast_node>(number_expr{val});
  // }

} // namespace dao
