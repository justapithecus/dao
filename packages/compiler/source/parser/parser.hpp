#pragma once

#include <iostream>
#include <iterator>
#include <memory>
#include <type_traits>
#include <vector>

#include "../ast/index.hpp"
#include "token.hpp"

namespace dao {

  using token_cursor = std::vector<token>::const_iterator;

  class parse_context {
    std::vector<token> const &tokens_;
    token_cursor              cursor_;

    // TODO(andrew): errors

  public:
    explicit parse_context(std::vector<token> const &tokens);

    [[nodiscard]]
    auto is_eof() const;

    [[nodiscard]]
    auto peek() const;

    /// seek
    ///
    /// Seeks and returns a cursor to the newly advanced position
    auto seek();

    /// eat
    ///
    /// Eats and returns the value of that token
    [[nodiscard]]
    auto eat();

    /// rewind
    ///
    /// Rewinds and returns a cursor to its previous position
    auto rewind();

    /// skip
    ///
    /// Skips consecutive ineffectual new-line tokens
    auto skip();
  };

  class parser {
    parse_context ctx_;

  public:
    explicit parser(std::vector<token> const &tokens);

    /// Parses tokens into an abstract syntax tree
    ///
    /// <program_ast> ::= { <expr> }
    auto parse() -> ast;

  private:
    /// Parses a literal
    ///
    /// <digit> ::= [0-9]
    /// <numeral_expr> ::= { <digit> }
    ///
    /// <string_literal> ::= '"' { <letter> | <digit> | <special_char> } '"'
    ///
    /// <literal> ::= <numeral_expr> | <string_literal>
    template <typename T>
    auto parse_literal() -> ast_node;

    /// Parses any expression
    ///
    /// <expr> ::= { <external_linkage> | <function_def> | <primary_expr> }
    auto parse_expr() -> ast_node;

    /// Parses a primary expression
    ///
    /// <primary_expr> ::= <identifier_expr> | <numeral_expr> | <parenthetical_expr> | <binary_expr>
    auto parse_primary_expr() -> ast_node;

    /// Parses a simple identifier expression
    ///
    /// <letter> ::= [A-Za-z]
    /// <digit> ::= [0-9]
    /// <identifier_expr> ::= <letter> { <letter> | <digit> | '_' }
    auto parse_identifier_expr() -> ast_node;

    /// Parses a binary expression, i.e. an expression that evaluates two expressions using
    /// a binary operator.
    ///
    ///  TODO(andrew): support more binary operators
    /// <binary_op> ::= { '*' | '+' | '-' | '/' | '>' | '<' }
    /// <binary_expr> ::= <primary_expr> { <binary_op> <primary_expr> }
    auto parse_binary_expr(std::uint8_t op_precedence = 0) -> ast_node;
    auto parse_binary_expr(ast_node lhs, std::uint8_t op_precedence = 0)
      -> ast_node;

    /// Parses a parenthetical expression, i.e. an expression group with highest evaluation
    /// precedence.
    ///
    /// <parenthetical_expr> ::= '(' <binary_expr> ')'
    auto parse_parenthetical_expr() -> ast_node;

    /// Parses a function argument
    ///
    /// TODO(andrew): add type hints
    /// <function_arg> ::= <identifier>
    auto parse_function_arg() -> function_arg;

    /// Parses a function argument sequence
    ///
    /// <function_arg_seq_multi> ::= <function_arg> | ( <function_arg> ',' <function_arg_seq_multi> )
    /// <function_arg_seq> ::= <empty> | <function_arg_seq_multi>
    auto parse_function_arg_seq() -> std::vector<function_arg>;

    /// Parses a function prototype
    ///
    /// <function_arg> ::= <identifier_seq>
    /// <function_proto> ::= 'function' <identifier_expr> '(' <function_arg_seq> ')'
    auto parse_function_proto() -> function_proto;

    /// Parses a function definition
    ///
    /// <function_def> ::= <function_proto> <primary_expr>
    auto parse_function_def() -> ast_node;

    /// Parses an expression sequence
    ///
    /// <expr_seq_multi> ::= <primary_expr> | ( <primary_expr> ',' <expr_seq_multi> )
    /// <expr_seq> ::= <empty> | <expr_seq_multi>
    auto parse_expr_seq() -> std::vector<ast_node>;

    /// Parses a function call
    ///
    /// <function_call> ::= <identifier_expr> '(' <expr_seq> ')'
    auto parse_function_call(std::string callee) -> ast_node;

    /// Parses an external linkage declaration
    ///
    /// <external_linkage_kind> ::= 'C'
    /// <external_linkage> ::= 'external' '(' <external_linkage_kind> ')' <function_proto>
    auto parse_external_linkage() -> ast_node;

    /// Parses an if/then/else expression
    ///
    /// <else_expr> ::= 'else' <primary_expr>
    /// <if_expr> ::= 'if' <primary_expr> 'then' { <expr> } [ <else_expr> ]
    auto parse_if_expr() -> ast_node;

    /// Parses a type alias
    ///
    /// <type_alias> ::= 'alias' <identifier_expr> 'as' <identifier_expr>
    auto parse_type_alias() -> ast_node;
  };

} // namespace dao
