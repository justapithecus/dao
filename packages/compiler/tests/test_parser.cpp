#include "test_utils.hpp"

#include "parser/parser.hpp"
#include "parser/state_machine.h"

auto directory{Approvals::useApprovalsSubdirectory("golden-files")};

auto main() -> int {

  "identifier_expression_simple"_test = [] {
    auto tokens{load_tokens("identifier_expression_simple.json")};
    auto ast{dao::parse(tokens)};

    json_writer writer{std::move(ast)};
    Approvals::verify(writer);
  };

  "numeral_expression"_test = [] {
    auto tokens{load_tokens("numeral_expression.json")};
    auto ast{dao::parse(tokens)};

    json_writer writer{std::move(ast)};
    Approvals::verify(writer);
  };

  "binary_expression_simple"_test = [] {
    auto tokens{load_tokens("binary-expressions/simple.json")};
    auto ast{dao::parse(tokens)};

    json_writer writer{std::move(ast)};
    Approvals::verify(writer);
  };

  "binary_expression_operator_precedence_1"_test = [] {
    auto tokens{load_tokens("binary-expressions/operator_precedence_1.json")};
    auto ast{dao::parse(tokens)};

    json_writer writer{std::move(ast)};
    Approvals::verify(writer);
  };

  "parenthetical_expression"_test = [] {
    auto tokens{load_tokens("parenthetical_expression.json")};
    auto ast{dao::parse(tokens)};

    json_writer writer{std::move(ast)};
    Approvals::verify(writer);
  };

  "binary_expression_operator_precedence_2"_test = [] {
    auto tokens{load_tokens("binary-expressions/operator_precedence_2.json")};
    auto ast{dao::parse(tokens)};

    json_writer writer{std::move(ast)};
    Approvals::verify(writer);
  };

  "function_prototype"_test = [] {
    auto tokens{load_tokens("functions/prototype.json")};
    auto ast{dao::parse(tokens)};

    json_writer writer{std::move(ast)};
    Approvals::verify(writer);
  };

  "function_definition"_test = [] {
    auto tokens{load_tokens("functions/definition.json")};
    auto ast{dao::parse(tokens)};

    json_writer writer{std::move(ast)};
    Approvals::verify(writer);
  };

  "function_call"_test = [] {
    auto tokens{load_tokens("functions/call.json")};
    auto ast{dao::parse(tokens)};

    json_writer writer{std::move(ast)};
    Approvals::verify(writer);
  };
}
