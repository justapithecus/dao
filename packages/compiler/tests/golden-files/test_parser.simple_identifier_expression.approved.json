{
  "root": {
    "type": "identifier_expr",
    "value": "function"
  }
}
